core/types/client/opts.go

```diff
@@ -45,13 +49,18 @@ func (c *Options) Apply(opts *Options) {
 		c.ChainID = opts.ChainID
 	}
 
+	if opts.Conn != nil {
```

Comment:
Just a thought on the style.

TBH, I don't like this way of passing options, i.e. `Client(xx, optionObj)`, I prefer `Client(xx, WithOptA, WithOptB...)`, and I kind of want to be able to apply those options to all our clients.
Currently we can reuse options though embed(like core/gatewayclient/opts.GatewayOptions`) , I'm not sure how to do it in `Client(xx, WithOptA...)` style, maybe using interface.

Reply:
I don't have strong feelings about this, but the functional options pattern is generally more cumbersome and time consuming to use compared to a config struct.  Yet most of our code does functional options, so might as well here.

In general, I think the main benefit of functional options is on the highest level exported APIs because a config struct changing is technically a breaking change to the constructor while new option functions are not.   With how unstable our Go APIs are all around, this benefit is almost non-existent.

Ironically, the most benefit for functional options is in `core/client.NewClient`, but the transport-aware clients deeper in core/rpc/client/... benefit less from doing it that way.

Anyway, I don't mind any options refactoring you'd like to do, but the more internal things get, the less important it is to use functional options.

Reply:
Yeah, just have this thought so it's consistent everywhere.  
I can work with current approach, at least all exposed clients are using config struct, it's consistent.